contracts/src/KyberElasticMultiQuoter.sol

```diff
@@ -0,0 +1,316 @@
+
+            // NOTE: this is equivalent to UniswapV3's zeroForOne.
+            // if tokenIn < tokenOut, token input and specified token is token0, swap from 0 to 1
+            bool isToken0 = tokenIn < tokenOut;
+            IPool pool = IPool(factory.getPool(tokenIn, tokenOut, fee));
+
+            // multiswap only accepts int256[] for input amounts
+            int256[] memory amounts = new int256[](amountsIn.length);
+            for (uint256 i = 0; i < amountsIn.length; ++i) {
+                amounts[i] = int256(amountsIn[i]);
+            }
+
+            MultiSwapResult memory result = multiswap(
+                pool,
+                isToken0,
+                amounts,
+                isToken0 ? TickMath.MIN_SQRT_RATIO + 1 : TickMath.MAX_SQRT_RATIO - 1
+            );
+
+            for (uint256 i = 0; i < amountsIn.length; ++i) {
+                amountsIn[i] = isToken0 ? uint256(-result.amounts1[i]) : uint256(-result.amounts0[i]);
+                gasEstimate[i] += result.gasEstimates[i];
+            }
+
+            // decide whether to continue or terminate
+            if (PathHelper.hasMultiplePools(path)) {
+                path = PathHelper.skipToken(path);
+            } else {
+                return (amountsIn, gasEstimate);
+            }
+        }
+    }
+
+    // TODO: same as uniswapv3 multiquoter logic. see if we can make generic.
+    function quoteExactMultiOutput(
+        IFactory factory,
+        bytes memory path,
+        uint256[] memory amountsOut
+    ) public view returns (uint256[] memory amountsIn, uint256[] memory gasEstimate) {
+        for (uint256 i = 1; i < amountsOut.length; ++i) {
+            require(amountsOut[i] >= amountsOut[i - 1], "multiswap amounts must be monotonically inreasing");
+        }
+        gasEstimate = new uint256[](amountsOut.length);
+        while (true) {
+            (address tokenOut, address tokenIn, uint24 fee) = PathHelper.decodeFirstPool(path);
+
+            // NOTE: this is equivalent to UniswapV3's zeroForOne.
+            // if tokenIn > tokenOut, output token and specified token is token0, swap from token1 to token0
+            bool isToken0 = tokenIn > tokenOut;
+            IPool pool = IPool(factory.getPool(tokenIn, tokenOut, fee));
+
+            // multiswap only accepts int256[] for input amounts
+            int256[] memory amounts = new int256[](amountsOut.length);
+            for (uint256 i = 0; i < amountsOut.length; ++i) {
+                amounts[i] = -int256(amountsOut[i]);
+            }
+
+            MultiSwapResult memory result = multiswap(
+                pool,
+                isToken0,
+                amounts,
+                isToken0 ? TickMath.MAX_SQRT_RATIO - 1 : TickMath.MIN_SQRT_RATIO + 1
+            );
+
+            for (uint256 i = 0; i < amountsOut.length; ++i) {
+                amountsOut[i] = isToken0 ? uint256(result.amounts0[i]) : uint256(result.amounts1[i]);
```

Comment:
Do we need to check that the deltas for amount received lines up with `amounts[i]`?

Reply:
good catch -- let me address this in a future PR. i'll probably run into the same issue you did with Uniswap.